pkg/operation/etcdencryption/etcdencryptioninfodata.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		return nil, err
+	}
+	if infoData != nil {
+		etcdEncryptionState, ok := infoData.(*ETCDEncryptionState)
+		if !ok {
+			return nil, fmt.Errorf("could not convert GardenerResourceData entry %s to ETCDEncryptionState", common.ETCDSecretsEncryptionStateName)
+		}
+
+		return etcdEncryptionState, nil
+	}
+	return nil, nil
+}
+
+// GetETCDEncryptionKey retrieves the ETCDEncryptionKey from the gardenerResourceDataList.
+func GetETCDEncryptionKey(gardenerResourceDataList gardencorev1alpha1helper.GardenerResourceDataList) (*ETCDEncryptionKey, error) {
+	infoData, err := infodata.GetInfoData(gardenerResourceDataList, common.ETCDSecretsEncryptionKeyDataName)
+	if err != nil {
+		return nil, err
+	}
+	if infoData != nil {
+		encryptionKey, ok := infoData.(*ETCDEncryptionKey)
+		if !ok {
+			return nil, fmt.Errorf("could not convert GardenerResourceData entry %s to ETCDEncryptionKey", common.ETCDSecretsEncryptionKeyDataName)
+		}
+		return encryptionKey, nil
+	}
+	return nil, nil
+}
+
+// UpsertETCDDataInShootState upserts the provided ETCDEncryptionKey and ETCDEncryptionState if they are not nil into the gardenerResourceList.
+func UpsertETCDDataInShootState(gardenerResourceList *gardencorev1alpha1helper.GardenerResourceDataList, etcdEncryptionKey *ETCDEncryptionKey, etcdEncryptionState *ETCDEncryptionState) error {
+	if etcdEncryptionKey != nil {
+		if err := infodata.UpsertInfoData(gardenerResourceList, common.ETCDSecretsEncryptionKeyDataName, etcdEncryptionKey); err != nil {
+			return err
+		}
+	}
+	if etcdEncryptionState != nil {
+		if err := infodata.UpsertInfoData(gardenerResourceList, common.ETCDSecretsEncryptionStateName, etcdEncryptionState); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GenerateOrSyncEncryptionKeyFromSecret generates a new ETCDEncryptionKey if the provided secret is nil, otherwise it fetches the required data from it.
+func GenerateOrSyncEncryptionKeyFromSecret(secret *corev1.Secret) (*ETCDEncryptionKey, error) {
+	if secret != nil {
+		conf, err := ReadSecret(secret)
+		if err != nil {
+			return nil, err
+		}
+		name, key, err := GetSecretKeyForResources(conf, common.EtcdEncryptionEncryptedResourceSecrets)
+		if err != nil {
+			return nil, err
+		}
+		etcdKey := &ETCDEncryptionKey{
+			Key:  key,
+			Name: name,
+		}
+		return etcdKey, nil
+	}
+
+	key, err := NewEncryptionKey(time.Now(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	etcdKey := &ETCDEncryptionKey{
+		Key:  key.Secret,
+		Name: key.Name,
+	}
+	return etcdKey, nil
+}
+
+// GenerateOrSyncEncryptionStateFromSecret syncs the ETCDEncryptionState data from the provided secret if it is not nil and the gardenerResourceDataList.
+// Otherwise, it generates a new ETCDEncryptionState.
+func GenerateOrSyncEncryptionStateFromSecret(gardenerResourceList gardencorev1alpha1helper.GardenerResourceDataList, secret *corev1.Secret) (*ETCDEncryptionState, error) {
```

Comment:
I found it confusing that this is generated from secret. Actually, you only use the secret to detect whether "force plain text" is enforced. Can you pass this information to the function and rename it?

```suggestion
func GenerateOrSyncEncryptionState(gardenerResourceList gardencorev1alpha1helper.GardenerResourceDataList, forcePlaintextSecrets bool) (*ETCDEncryptionState, error) {
```